File: legacy/mobile/lib/core/protocol_constants.dart
/// BLE/input protocol constants, mirrored from docs/protocol.md.
library;

abstract final class ProtocolConst {
  // UUIDs
  static const String serviceUuid = 'd8e6f9a0-4000-4000-8000-000000000001';
  static const String txUuid = 'd8e6f9a0-4000-4000-8000-000000000101'; // phone -> PC (notify)
  static const String rxUuid = 'd8e6f9a0-4000-4000-8000-000000000102'; // PC -> phone (write)

  static const int version = 0x01;

  // Header flags
  static const int flagAckRequested = 0x01;

  // Packet types
  static const int typeHello = 0x01;
  static const int typeHelloAck = 0x02;
  static const int typePing = 0x03;
  static const int typePong = 0x04;
  static const int typeDisconnect = 0x05;
  static const int typeMouseMove = 0x10;
  static const int typeMouseButton = 0x11;
  static const int typeMouseScroll = 0x12;
  static const int typeMouseNav = 0x13;
  static const int typeJoystickAxis = 0x30;

  // MOUSE_MOVE payload [dx i16 @0][dy i16 @2][buttons @4][flags @5]
  // MOUSE_SCROLL payload [dx i16 @0][dy i16 @2][flags @4]
  // MOUSE_BUTTON payload [button @0][pressed @1]
  // MOUSE_NAV   payload [action @0]
  // DISCONNECT  payload [reason @0]

  // Button mask
  static const int btnLeft = 0x01;
  static const int btnRight = 0x02;
  static const int btnMiddle = 0x04;
  static const int btnBack = 0x08;
  static const int btnForward = 0x10;

  // Scroll flags
  static const int scrollHorizontalAxis = 0x01;

  // Nav actions (matches receiver NavAction)
  static const int navBack = 0x00;
  static const int navForward = 0x01;
  static const int navPrevTrack = 0x02;
  static const int navNextTrack = 0x03;

  /// Max payload bytes fitting a 20-byte ATT payload with the 6-byte header.
  static const int maxPayload = 14;

  /// Peer considered lost after no PONG for this long.
  static const Duration pingTimeout = Duration(seconds: 10);

  static const Duration pingInterval = Duration(seconds: 5);
}

/// Nav actions (indices match receiver NavAction enum values 0x00..0x03).
enum NavAction { back, forward, prevTrack, nextTrack }

/// Small typed wrapper the protocol layer mints; packets themselves are
/// uint8 arrays so IO stays zero-copy-simple.
final class MouseMoveData {
  const MouseMoveData(this.dx, this.dy, this.buttons);
  final int dx;
  final int dy;
  final int buttons;
}

File: legacy/windows-receiver/src/Protocol/ProtocolConstants.cs
namespace Dtcon.Receiver.Protocol;

/// <summary>
/// UUID registry and numeric protocol constants. Single source of truth in code;
/// the durable spec is docs/protocol.md and shared/protocol/protocol.md.
/// </summary>
public static class ProtocolConstants
{
    public const byte Version = 0x01;

    /// <summary>Bytes of the fixed 6-byte packet header (version, type, flags, sequence, length LE).</summary>
    public const int HeaderSize = 6;

    /// <summary>Largest payload that fits in one BLE notification at the default 23-byte MTU.</summary>
    public const int MaxPayload = 14;

    // UUIDs
    public const string ServiceUuid = "d8e6f9a0-4000-4000-8000-000000000001";
    public const string TxUuid = "d8e6f9a0-4000-4000-8000-000000000101"; // phone -> PC (notify)
    public const string RxUuid = "d8e6f9a0-4000-4000-8000-000000000102"; // PC -> phone (write)

    // Flags bit 0x01 = ACK requested.
    public const byte FlagAckRequested = 0x01;
}

public enum PacketType : byte
{
    Hello = 0x01,
    HelloAck = 0x02,
    Ping = 0x03,
    Pong = 0x04,
    Disconnect = 0x05,
    MouseMove = 0x10,
    MouseButton = 0x11,
    MouseScroll = 0x12,
    MouseNav = 0x13,
    // Reserved for joystick: JoystickAxis = 0x30, JoystickButton = 0x31
}

/// <summary>Protocol button bits (wire format). Values match the spec's button mask.</summary>
public enum MouseButtonId : byte
{
    None = 0x00,
    Left = 0x01,
    Right = 0x02,
    Middle = 0x04,
    Back = 0x08,
    Forward = 0x10,
}

public enum NavAction : byte
{
    Back = 0,
    Forward = 1,
    Previous = 2,
    Next = 3,
}

public enum DisconnectReason : byte
{
    User = 0,
    TransportError = 1,
    Timeout = 2,
}